spec/models/ability_spec.rb

```diff
@@ -89,11 +89,104 @@
+
+        context 'when user has already registered' do
+          let(:user_registration) { create(:registration, conference: conference, user: user) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: user_registration.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: user_registration.conference_id)) }
+        end
+
+        context 'when user has not registered and registrations are open with no registration_limit_exceeded' do
+          let(:conference_with_open_registration_and_limit_not_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
+
+          it{ should be_able_to(:new, Registration.new(conference: conference_with_open_registration_and_limit_not_excedeed)) }
+          it{ should be_able_to(:create, Registration.new(conference: conference_with_open_registration_and_limit_not_excedeed)) }
+        end
+
+        context 'when user has not registered and registrations are open with registration_limit_exceeded' do
+          let(:conference_with_open_registration_and_limit_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
+          let(:registration_for_user_with_open_registration_and_limit_excedeed) { create(:registration, conference: conference_with_open_registration_and_limit_excedeed, user: user2) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+        end
+      end
+
+      context 'when user is a speaker with event confirmed' do
+        let(:conference_with_speaker_confirmed) { create(:conference) }
+        let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
+
+        before :each do
+          event_with_speaker_confirmed.speakers << user
+        end
+
+        context 'when speaker has already registered' do
+
+          before :each do
+            create(:registration, conference: conference_with_speaker_confirmed, user: user2)
+          end
+
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed)) }
+        end
+
+        context 'And the registration period is set' do
+          let(:registration_period) { create(:registration_period) }
```

Comment:
In these tests, I would always expect the registration_period to be associated with a relevant conference immediately. It is a bit confusing how you later on associate it with a different conference.

Think about it logically, what do you do first if you are a user in OSEM?
Don't you first create a new conference, and then go setup a registration_period (for that specific conference)?
If you keep this logical order, it is way easier to read the tests later on.

This line can simply be
create(:registration_period, conference: conference_with_speaker_confirmed)

Reply:
I associate a conference with a different conference because value will be cached
across multiple calls in the same example but not across examples. So when there is another example test it won't be affect by the previous or any other test.